### ISSUE TITLE
grub support for zfs encryption

### ISSUE THREAD

**Issue by ptx0** [zfsonlinux/zfs#7153]
grub2 does not recognize the root filesystem on a pool with `encryption@enabled`

### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Gentoo
Distribution Version    | 17.0
Linux Kernel                 | 4.14.15
Architecture                 | amd64
ZFS Version                  | master
SPL Version                  | master

### Describe the problem you're observing
Enabling encryption makes a pool unable to boot.

### Describe how to reproduce the problem
1. Create a pool and install OS to encrypted volume
2. Run `grub-probe -vvv /`
Output: "error: unknown filesystem"
Expected: "zfs"

It's notable that there is already a zfscrypt module in grub upstream:
http://git.savannah.gnu.org/cgit/grub.git/log/grub-core/fs/zfs/zfscrypt.c

**Comment by bunder2015** [comment]
A few of us had a huge long discussion on IRC, but posting here for the sake of visibility.  Grub hasn't had any ZFS features added to it in quite some time.  I had attempted to add a few myself last year and got shouted down by two or three developers over licensing.  Personally, I've moved on to EFI.  That said, it would be nice to have more features added to grub, but I'm not going to hold my breath, and I don't envy the person who picks up this torch.

**Comment by lundman** [comment]
I was considering adding openzfs crypt support to the work I did over at u-boot, once the crypto support lands (on Illumos). Sounds like there isn't much help to be expected from grub guys?

**Comment by loli10K** [comment]
Is this a ZFS issue or a GRUB issue (duplicate of https://github.com/zfsonlinux/grub/issues/24)?

**Comment by bunder2015** [comment]
Is that repo for all distros, or just deb/ubu variants?  I can't tell.

**Comment by behlendorf** [comment]
In fact the repository is quite old.  We should probably migrate the outstanding issues which are still relevant to either the main ZFS issue tracker or as upstream grub issues, then retire the repository.

**Comment by ptx0** [comment]
if it ends up as an upstream grub issue it will just die without any attention given to it. since the code came from us, we should either fix it or provide an alternative (porting FreeBSD/illumos btx loader is probably the most straightforward solution here)

**Comment by rlaager** [comment]
In my experience, nobody at upstream GRUB cares much about ZFS support. They will accept patches from those who are interested, but such patches need to be GPL-licensed, not CDDL-licensed, so you can't copy anything from OpenZFS (unless all authors of that code agree to dual-license it).

More importantly, if the patch is non-trivial (which encryption support certainly would be), GPL-licensed is not sufficient: the original copyright holder must _assign the copyright_ (on all code contributed to GRUB) to the FSF. The paperwork on this is straightforward to do, but does represent an additional hurdle. The issue of whether the original copyright holder wants to do such a thing is much more complicated. So, in practice, it's best if the GRUB work is a re-implementation which is assigned to the FSF.

I did the copyright assignment for contributing code to GRUB, so I know what it entails. I also had to get a copyright release from my employer. My employer is small, so this was easy. It may not be easy (or even reasonably possible) at a larger company.

For Ubuntu, my current intention is to follow the current approach of unencrypted /boot, which means no GRUB support is necessary. Then the initramfs will unlock the filesystems. With ZFS native encryption, this would preferably be a ZFS /boot dataset in the root pool, as opposed to ext2 /boot like I'm using now in the HOWTO with LUKS.

**Comment by ptx0** [comment]
using an ext2 /boot means the system will default to the cfq scheduler on most systems, causing I/O timeouts for application when running scrub or send/recv or even high  load on a zvol. using noop for the entire drive whenever ZFS is running on it should be the default behaviour, on linux, if this is what we want to recommend to users. you also, unfortunately, don't get the benefits of ZFS checksum, mirroring, raid, or snapshots on the directory.

FreeBSD uses bpool and rpool where bpool is the unencrypted boot pool but on Linux we don't set the scheduler properly so that won't work, and it's a pain in the ass to deal with sizing. if I want to keep 25 kernels for testing but  originally intended to hold onto just a few, it would be inconvenient to say the least.

if we port btx from freebsd/illumos then we'll have the benefit of OpenZFS contributions as well as completely avoiding licensing and copyright concerns. distribution support for btx is not difficult considering Debian/kFreeBSD and gentoo/alt have already (had, or have) support for it in their userland.

**Comment by rlaager** [comment]
As I said, my intention is to use a ZFS dataset for /boot and not ext2.

I absolutely agree that ZFS for /boot is better, for checksumming, mirroring/raidz, and snapshots. Either way, the elevator is noop. The initramfs script in Ubuntu was patched (by me) to set elevator=noop on the partitions used in the root pool (since ZoL only sets elevator=noop for wholedisk pools). Unfortunately, this was lost when Debian (and thus Ubuntu) switched to upstream ZoL's initramfs script. But I got this change upstreamed into ZoL's initramfs script a release or two ago, so that is resolved again. But yeah, ZFS /boot is still better than ext2 /boot!

**Comment by kpande** [comment]
jfyi i've been using a v5000 pool that is restricted to features that are supported by grub, in order to boot from ZFS native encrypted rootfs. you can also use a v28 pool for extreme compatibility but it will be very slow.

**Comment by rlaager** [comment]
@kpande Are you doing an unencrypted /boot on ZFS? I assumed that GRUB would bail if native encryption was enabled on the pool at all, regardless of whether /boot is encrypted. If GRUB is happy with unencrypted /boot on ZFS w/ native encryption on that pool, that's great.

**Comment by kpande** [comment]
no, two separate pools, a feature-restricted one for /boot (similar to FreeBSD)

**Comment by kpande** [comment]
it should be trivial to modify grub to not reject the encryption feature but simply not attempt to access any encrypted filesystems directly. then you *could* boot directly from an unencrypted filesystem on a pool with encryption enabled.

**Comment by vozhyk-** [comment]
I've made a [one-line change](https://gitlab.com/linux-be/grub/compare/4-use-kernel-detection...8-dont-reject-pools-with-encryption-active) to not reject pools with `feature@encryption=active` and could boot from an unencrypted dataset.
This is not a complete fix for a user, though - `grub-probe` also accepts encrypted datasets, and then booting from them fails with a confusing error message (`error: compression algorithm inherit not supported`).

**Comment by pstch** [comment]
@vozhyk- this change does make it much easier in some cases : it would be no longer necessary to bypass grub-probe when booting an encrypted pool (using an unencrypted non-ZFS /boot).

**Comment by cytrinox** [comment]
Is this still an issue? I want to create a single pool with tank/boot unencryted but tank/master/{root,home,var} being encrypted. Will grub reject this pool?
The one-line change posted by @vozhyk- has gone. Is this patch now included in grub?

**Comment by kpande** [comment]
yes it is an issue, grub will reject it. the one line change is likely gone because it was not properly implemented. no fixes have made it into grub.

**Comment by vozhyk-** [comment]
@cytrinox I've updated the link to the change. The setup you mention should work with it, if you didn't activate feature flags not supported by GRUB other than encryption.

**Comment by misterhsp** [comment]
What's the module here for?

ls -l /boot/grub/x86_64-efi/zfscrypt.mod 
-rw-r--r-- 1 root root 8424 Mär 15 08:43 /boot/grub/x86_64-efi/zfscrypt.mod

**Comment by kpande** [comment]
unrelated module for oracle zfs encryption support in grub.

**Comment by sotiris-bos** [comment]
So, the only way to boot a kernel on a ZFS unencrypted dataset, on a ZFS pool with encryption enabled is to use the "8-dont-reject-pools-with-encryption-active" GRUB branch with suitable flags enabled on the pool?

What about the possibility of using rEFInd with a third party ZFS EFI driver?

**Comment by mskarbek** [comment]
> So, the only way to boot a kernel on a ZFS unencrypted dataset, on a ZFS pool with encryption enabled is to use the "8-dont-reject-pools-with-encryption-active" GRUB branch with suitable flags enabled on the pool?
> 
> What about the possibility of using rEFInd with a third party ZFS EFI driver?

Why would you need a third party EFI driver? If you plan to use EFI then you have to have ESP which is unencrypted and can contain kernel/initramfs with everything you need to boot from with encrypted pool. That works well with systemd-boot and some dracut magic.

**Comment by sotiris-bos** [comment]
> Why would you need a third party EFI driver? If you plan to use EFI then you have to have ESP which is unencrypted and can contain kernel/initramfs with everything you need to boot from with encrypted pool. That works well with systemd-boot and some dracut magic.

I am specifically saying that the kernel (and initramfs) is on an unencrypted ZFS dataset in this scenario. That means you would need an EFI partition for rEFInd and the EFI driver so that rEFInd can mount the unencrypted ZFS dataset that the kernel/initramfs resides and boot it.

@rlaager and @kpande where discussing above the possibility of having the kernel and initramfs on ZFS for: 

> checksumming, mirroring/raidz, and snapshots.
